lisa/tests/staging/schedutil.py

```diff
@@ -0,0 +1,271 @@
+        """
+        trace = self.trace
+        cpu = self.cpu
+
+        # schedutil_df also has a 'util' column that would conflict
+        schedutil_df = trace.df_events('schedutil_em')[['cpu', 'cost_margin']]
+        schedutil_df['from_schedutil'] = True
+
+        df_list = [
+            schedutil_df,
+            trace.analysis.load_tracking.df_cpus_signal('util'),
+            trace.analysis.load_tracking.df_cpus_signal('util_est_enqueued'),
+        ]
+
+        df = df_merge(df_list, filter_columns={'cpu': cpu})
+        df['from_schedutil'].fillna(value=False, inplace=True)
+        df.ffill(inplace=True)
+        df.dropna(inplace=True)
+
+        # Reconstitute how schedutil sees signals by subsampling the
+        # "master" dataframe, so we can look at signals coming from other
+        # dataframes
+        df = df[df['from_schedutil'] == True]
+        df.drop(columns=['from_schedutil'], inplace=True)
+
+        # If there are some NaN at the beginning, just drop some data rather
+        # than using fake data
+        df.dropna(inplace=True)
+
+        boost_points = (
+            # util_est_enqueued is the same as last freq update
+            (df['util_est_enqueued'].diff() == 0) &
+
+            # util_avg is increasing
+            (df['util'].diff() >= 0) &
+
+            # util_avg > util_est_enqueued
+            (df['util'] > df['util_est_enqueued'])
+        )
+        df['boost_points'] = boost_points
+
+        df['expected_cost_margin'] = (df['util'] - df['util_est_enqueued']).where(
+            cond=boost_points,
+            other=0,
+        )
+
+        # cost_margin values range from 0 to 1024
+        ENERGY_SCALE = 1024
+
+        for col in ('expected_cost_margin', 'cost_margin'):
+            df[col] *= 100 / ENERGY_SCALE
+
+        # We cannot know if the first row is supposed to be boosted or not
+        # because we lack history, so we just drop it
+        return df.iloc[1:]
+
+    def _plot_test_boost(self, df):
+        task = self.rtapp_tasks[0]
+        analysis = self.trace.analysis.frequency
+        fig, axis = analysis.setup_plot(interactive=False)
+        df['cost_margin'].plot(ax=axis, drawstyle='steps-post', color='r')
+        df['boost_points'].astype('int', copy=False).plot(ax=axis, drawstyle='steps-post', color='black')
+        df['expected_cost_margin'].plot(ax=axis, drawstyle='steps-post', color='blue')
+
+        self.trace.analysis.tasks.plot_task_activation(task, axis=axis, overlay=True)
+
+        axis.legend()
+        analysis.save_plot(fig, filepath=os.path.join(self.res_dir, 'ramp_boost.svg'))
+        return axis
+
+    @requires_events('sched_switch')
+    def trace_window(self, trace):
+        """
+        Skip the first phase in the trace, since it is heavily influenced by
+        what was immediately preceding it.
+        """
+        profile = self.rtapp_profile
+        phase_duration = max(
```

Comment:
This can probably be moved over to the new phase windows API. However, seeing as you've already tested the thing, I wouldn't mind if this was merged in staging as-is. We can then do a batch convert (misfit needs that treatment too).

Reply:
Yes, we will go over the tests and update them with the new API once merged